Send all logs to a remote location

hubertp opened the pull request:
### Pull Request Description

This change adds appender for a generic remote endpoint. The change extracted common parts between telemetry and general logging capabilities, to avoid code duplication. 

### Important Notes

This change removes Sentry appender - it's obsolete.
Removed `telemetry` as a dependency of `runtime-integration-tests` as it was not necessary.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.

hubertp committed:
Generalize Telemetry Appender to send all logs

This change refactors existing Telemetry logic, so that we now have
OpenSearch and Telemetry remote logging, both using _mostly_ the same
logic.
The split comes from the fact that OpenSearch appender requires a bit
more tweaks to its configuration. For example, we want to avoid an
infinite loop of logging when connection failures occur.

hubertp committed:
Add missing files

hubertp committed:
fmt

hubertp committed:
Reducing the amount of noise coming from logs

hubertp committed:
drop sentry

hubertp committed:
mdc

Akirathan commented on lib/scala/project-manager/src/main/resources/application.conf:
```diff
@@ -73,12 +79,22 @@ logging-service {
-          dsn = "<dsn>"
-          dsn = ${?ENSO_APPENDER_SENTRY_DSN}
+          name = "opensearch",
+          uri = "https://api.ensoanalytics.com"
```
Is this a correct production URI?

hubertp replied:
It will be something like that. I'd like to get rid of the hardcoded url which is not very Enso'ish.

Akirathan replied:
Removing hardcoded URLs is tracked in https://github.com/enso-org/enso/issues/12564

hubertp replied:
Yes, might as well do it partially here. This requires support from cloud, however.

Akirathan reviewed (commented):
Looks OK in general. Would be nice to separate the unrelated removal of sentry and other logging cleanups into a separate PR. Would be also nice to have some unit tests for opensearch appender.

hubertp committed:
Split telemetry into common and opensearch logic

Refactored initial Telemetry Appender logic so that we have common
remote logging logic separated from the implementation.
That way generic logging can use most of the logic.

hubertp committed:
Remove telemetry from runtime-integratio-tests

Looks like `TestLogFormatter` shouldn't be in
`runtime-integration-tests` in the first place.
The change also simplified sbt setup.

hubertp committed:
s/logging.service.remote/logging.service

hubertp committed:
reducing verbosity

hubertp committed:
s/telemetry.test/telemetry

hubertp committed:
Add OpenSearchFormatter tests

hubertp committed:
fmt

JaroslavTulach commented on build.sbt:
```diff
@@ -592,14 +594,6 @@ val zio = Seq(
-// === Sentry =================================================================
-
-val ioSentryVersion = "6.28.0"
-val ioSentry = Seq(
```
Good to see removing (long time unused) sentry support.

JaroslavTulach commented on build.sbt:
```diff
@@ -1827,6 +1872,7 @@ lazy val `project-manager` = (project in file("lib/scala/project-manager"))
   .dependsOn(pkg)
   .dependsOn(`logging-service-logback` % Runtime)
   .dependsOn(`logging-service-telemetry` % Runtime)
+  .dependsOn(`logging-service-opensearch` % Runtime)
```
`project-manager` has a _runtime only_ dependency on `logging-service-opensearch`. That looks like a fine _loosely coupling_.

JaroslavTulach commented on lib/java/logging-service-opensearch/src/main/module-info.java:
```diff
@@ -0,0 +1,15 @@
+  requires org.slf4j;
+  requires static org.openide.util.lookup.RELEASE180;
+
+  provides org.enso.logging.service.logback.RemoteAppender with
```
Someone will ask for `RemoteAppender` and obtain implementation of `OpenSearchAppender`.

JaroslavTulach commented on lib/scala/logging-service-logback/src/main/java/org/enso/logging/service/logback/OpenSearchAppender.java:
```diff
@@ -0,0 +1,17 @@
+    var loader = ServiceLoader.load(RemoteAppender.class, RemoteAppender.class.getClassLoader());
+    var appender =
+        loader.stream()
+            .filter(l -> l.type().getName().contains("OpenSearch"))
```
Searching for a class name is ugly. Some might say this is even more _loosely coupled_, but it is ugly.

JaroslavTulach commented on lib/java/logging-service-opensearch/src/main/java/org/enso/logging/service/opensearch/OpenSearchAppender.java:
```diff
@@ -0,0 +1,33 @@
+
+@org.openide.util.lookup.ServiceProvider(
+    service = org.enso.logging.service.logback.RemoteAppender.class)
+public final class OpenSearchAppender extends org.enso.logging.service.RemoteAppender {
```
Not sure what witticism to use? Maybe DRY? Why do we thing it make sense to **have three classes of the same name** in the system?

![obrazek](https://github.com/user-attachments/assets/5603245c-b256-468d-934a-ae6a3a2e389b)

This kind of _use the same name three times for three different things_ confuses me every time I try to approach this part of code base.

hubertp replied:
The separation allows us to not tightly couple with a particular appender in runtime deps. This is following the same pattern as [Telemetry Appender PR](#12488) and you didn't voice concerns about it back then.

JaroslavTulach reviewed (commented):
- I don't like using `OpenSearchAppender` as a name for three(!) different classes.
- I doubt such _"unification of names"_ helps clarity
- if nothing else, I have little idea during review what particular use of `OpenSearchAppender` refers to

Akirathan reviewed (commented):
Having two separate modules for telemetry and opensearch appenders is good. Couple of notes about module declarations:
- `org.enso.logging.service.opensearch` module is not in `componentModulePaths`  in `build.sbt` . This means that it will not get copied into `built-distribution/**/component`  directory, so that it will not be present on module-path. How can it work? Is that intentional?
- `org.enso.logging.service.telemetry`  module should have `requires org.enso.logging.service.remote` .
- Since all the service providers are declared in module descriptors, we probably don't need the `org.openide.util.lookup`  from NetBeans. Can be done in a separate PR.
- `org.enso.logging.service.logback`  module descriptor has `uses org.enso.logging.service.logback.TelemetryAppender` . That seems like a leftover. Instead, there should be `uses org.enso.logging.service.logback.RemoteAppender` .

Akirathan commented:
Are you planning to add unit tests for OpenSerachAppender similar to [TestTelemetry](https://github.com/enso-org/enso/blob/47cc5a5e18f54869b64e664ccea2d694b6bd7835/lib/java/logging-service-telemetry/src/test/java/org/enso/logging/service/telemetry/TestTelemetry.java) that uses the `http-test-helper` mock server?

JaroslavTulach commented:
+1 to replace `Lookup` by requires+provides. GraalVM24.2 has all the necessary fixes.

hubertp committed:
Merge branch 'develop' into wip/hubert/8517-send-remote-logs

hubertp commented:
> Are you planning to add unit tests for OpenSerachAppender similar to [TestTelemetry](https://github.com/enso-org/enso/blob/47cc5a5e18f54869b64e664ccea2d694b6bd7835/lib/java/logging-service-telemetry/src/test/java/org/enso/logging/service/telemetry/TestTelemetry.java) that uses the `http-test-helper` mock server?

I wasn't. Testing of log formatting seemed sufficient and adding a test that would involve such a mock server seemed quite pointless. It was already covered by telemetry and both were using the same infrastructure.

hubertp commented:
> Having two separate modules for telemetry and opensearch appenders is good. Couple of notes about module declarations:
> 
> * `org.enso.logging.service.opensearch` module is not in `componentModulePaths`  in `build.sbt` . This means that it will not get copied into `built-distribution/**/component`  directory, so that it will not be present on module-path. How can it work? Is that intentional?

Neither does telemetry. Isn't it because telemetry was only packaged for `project-manager` and that's actually desired?

> * `org.enso.logging.service.telemetry`  module should have `requires org.enso.logging.service.remote` .
> * Since all the service providers are declared in module descriptors, we probably don't need the `org.openide.util.lookup`  from NetBeans. Can be done in a separate PR.
> * `org.enso.logging.service.logback`  module descriptor has `uses org.enso.logging.service.logback.TelemetryAppender` . That seems like a leftover. Instead, there should be `uses org.enso.logging.service.logback.RemoteAppender` .

Sure, I can take a look at the modules, if that feels like a better way to tackle it.

Akirathan commented:
> Neither does telemetry. Isn't it because telemetry was only packaged for project-manager and that's actually desired?

You are right. I forgot we assemble them into `project-manager.jar` fat jar. So `org.openide.lookup` is probably still needed, since all the `module-info.class` are thrown away from the fat jar anyway.

hubertp committed:
Address PR comments

Ditched confusing OpenSearch/Telemetry classes and moved loading of
implementations to a base class of remote appender. Fixed assembly so
that we can have multiple providers for remote appenders.

hubertp commented:
As discussed, unfortunately service lookup has to stay because `project-manager` is not modularized. We could probably do it but it's an extra effort with limited advantages at the moment. We could probably do it easier after switching to a build system that supports JPMS natively.

This has been tested and works quite well. I will wait with merge until we get a default, human-readable, API endpoint to point to from cloud team. But an approval would be appreciated as, I think, I have addressed most of the concerns.

hubertp committed:
Merge branch 'develop' into wip/hubert/8517-send-remote-logs

hubertp committed:
trigger build

hubertp committed:
Adapt to new URI
